### ISSUE TITLE
Non-null assertions with createAsync?

### ISSUE THREAD

**Issue by klequis** [klequis/solid-discussions#11]
**SilentRhetoric — Yesterday at 5:39 AM**
When using Solid Start, is there any better pattern than using non-null assertions to access properties of a createAsync return object?

https://github.com/solidjs/solid-start/blob/179500ffd6855f7248de7aa6f3672dc2bac773f2/examples/hackernews/src/routes/stories/%5Bid%5D.tsx#L18

**mdynnl — Yesterday at 5:44 AM**
Show with callback form

**SilentRhetoric — Yesterday at 5:45 AM**
I see that being used in the example, but it doesn't actually appear to remove the need to !. or ?. the properties access 

**Atila — Yesterday at 5:56 AM**
@SilentRhetoric <Show> w/ callback seems to work well around here:

![image](https://github.com/user-attachments/assets/76d76beb-74b7-42f0-86f3-95caa83cc7c2)

**SilentRhetoric — Yesterday at 5:57 AM**
Ohhhhhh
I have to ask the question, then

**Atila — Yesterday at 5:59 AM**
though because I have <Suspense> I think in practice that would be the same as
```
        <Suspense fallback={<div>Loading...</div>}>
          <>
            <div>{pokemons()![0].title}</div>
```

**SilentRhetoric — Yesterday at 6:00 AM**
Where should I find the example that implements these patterns?

**Atila — Yesterday at 6:02 AM**
I don't think they're described in the docs - we have been more focused in documenting all the APIs and the guides kind of fell in priority. But, I think they should live on: https://docs.solidjs.com/solid-start/building-your-application/data-loading#data-loading
[6:03 AM]
and we should probably have a Guide or Concepts entry pointing out the differences between <Suspense> and <Show> for such things
[6:04 AM]
TL;DR: Suspense will render both paths (the fallback and the elements, even without the data - but it will only attach the fallback to the DOM until data resolves.

Show will render/attach only the one that matches the conditions.

Once the condition changes, Show will render hte other one, dettach the current, and attach the new. While Suspense will just dettach/attach without needing to render things.

**SilentRhetoric — Yesterday at 6:06 AM**
First time using Solid Start to do SSR... so let me run this back to see if I understand...

**Atila — Yesterday at 6:07 AM**
So, using the <Show>  for toggling data is probably less appropriate than using <Suspense>.

As a rule-of-thumb, I'd prefer:

1. `<Suspense>`
2. `<Suspense>` + `deferStream` on `createAsync`
3. `<Suspense>` + `<Show>`

**SilentRhetoric — Yesterday at 6:09 AM**
How does this approach differ for loading on the client vs the server?
My page is almost exactly like the HackerNews example--just one API fetch to get some data which then gets displayed
I don't really need to use SSR for this app, but I am using it as an opportunity to learn the Solid Start + SSR patterns

**Atila — Yesterday at 6:11 AM**
for   we usually recommend creating a Server Function for all data loading. This creates a RPC  layer between your app and the server. This way the data-fetch roundtrip happens within the server boundaries, not relying on your user's connection.
SolidStart will also deduplicate the requests and the artifacts going down the wire, so the overhead is minimal for each rountrip.

Also, you can use the route preloading to trigger the fetch request before you navigate.

**SilentRhetoric — Yesterday at 6:11 AM**
I did that using Hackernews as a model

**Atila — Yesterday at 6:13 AM**
yeah, in that smaller app usage, the only benefit you'd get would be to fetching data together with rendering. So when the server replies the request, it will send the rendered HTML instead of using JS to render it all and then fetch the data

**SilentRhetoric — Yesterday at 6:13 AM**
Following mdynnl's advice, adding the Show with the callback function improved the ergonomics because then I don't have to worry about undefined within the Show
[6:14 AM]
So I've got your approach #3 now

**Atila — Yesterday at 6:14 AM**
Approach 2 will make no difference if you have only 1 route

**SilentRhetoric — Yesterday at 6:15 AM**
Yep

**Atila — Yesterday at 6:15 AM**
and then 1, in your case, I'd say it's more of a microoptimization (?)
but perhaps you just want to align with best practices (though it forces some type-assertions to appease the tsc overlords)

**SilentRhetoric — Yesterday at 6:16 AM**
We must keep TypeScript happy
Is the Suspense achieving anything in this SSR pattern?

**Jasmin — Yesterday at 6:18 AM**
But does TypeScript make us happy in return? ()

**SilentRhetoric — Yesterday at 6:18 AM**
Sometimes! Better to find errors at compile time

**Atila — Yesterday at 6:19 AM**
it allows concurrent rendering, so the rendering is done when the data arrives. 

It achieves way more once you get multiple requests/boundaries, or a more complex view.
Show causes a little DOM switcharoo that isn't ideal - but in a single-view app, with a single fetch call - that's not a significant overhead. Imho

**SilentRhetoric — Yesterday at 6:25 AM**
Wouldn't Show inside of a Suspense be the best combination, then?

Suspense will render both 
a. Fallback (e.g. "Loading...")
b. Actual page

but!

If there is a nested Show with no fallback, then Show would prevent b from rendering anything until the data arrives

And then Show with a callback improves the code ergonomics because you don't need to handle the undefined data case

**Hussein — Yesterday at 6:26 AM**
what does deferStream do?

**Atila — Yesterday at 6:26 AM**
the Show inside the Suspense would make the background rendering null... so there would be no point of a suspense boundary

**Atila — Yesterday at 6:27 AM**
it holds navigation until data arrives.

createAsync without deferStream => navigation happens and shows suspense fallback

createAsyncwith deferStream => navigation holds and doesn't show the route until the data arrives - suspense boundary/fallback is never triggered in the arriving route

deferStream = "defer streaming until the data is all there"

**SilentRhetoric — Yesterday at 6:28 AM**
Ohhhhhh

**Hussein — Yesterday at 6:28 AM**
that sounds bad to me. isn't it better to show static content (while waiting for data) instead of nothing?

**SilentRhetoric — Yesterday at 6:29 AM**
I think it depends what your page looks like without data

**Atila — Yesterday at 6:29 AM**
depends on your use-case. Let's say this is a protected route? You may not want to navigate until you know for sure the user has access

**SilentRhetoric — Yesterday at 6:29 AM**
And how smooth the transition is from no data --> data
I actually do have a bit of nesting going on, I just realized, because there's two images that can only begin loading once the page's initial data is fetched, which includes the urls for those images

**Atila — Yesterday at 6:32 AM**
and that's why createAsync is built for Suspense - you drop those <Show> and all requests happen in parallel
SolidStart is the waterfall destroyer

**SilentRhetoric — Yesterday at 6:33 AM**
On my page it goes:
Fetch JSON
2a. Load imageA with URL from JSON in 1
2b. Load imageB from URL from JSON in 1
So there is still some requirement for a waterfall, but I think the answer here is nested Suspense

**Atila — Yesterday at 6:34 AM**
I think that's a non-issue with SSR though

**SilentRhetoric — Yesterday at 6:35 AM**
OK so we're really getting to the meat of the issue now
Because the images are much slower to load than the JSON in 1
So I have to make a design decision here about if I want to render the page before those images load

**Hussein — Yesterday at 6:37 AM**
good example

**Hussein — Yesterday at 6:40 AM**
so is this like moving from renderToStream to renderToStringAsync per route?

**Atila — Today at 2:51 AM**
that's a great question. I think so.
renderToStream => concurrent - we send the data with the Suspense unresolved
renderToStrincAsync => sending the rendered artifact waits for the data to arrive and renders thing only once. I think that's fundamentally the same as createAsync(..., {deferStream: true}

maybe @lxsmnsyc 🤖 is the guy we need to confirm, because I'm not totally confident on the 1:1 relationship.

**lxsmnsyc 🤖 — Today at 2:52 AM**
yeah rTSA is basically just a blocked rTS
